src/shared/sleep-config.c

```diff
@@ -251,61 +251,55 @@ int fetch_batteries_capacity_by_name(Hashmap **ret) {
         const char *x;
         int r;
 
         assert(dev);
-        assert(property);
+        assert(attr);
         assert(state);
 
-        r = sd_device_get_property_value(dev, property, &x);
-        if (r == -ENOENT)
-               log_device_debug_errno(dev, r, "battery device property %s is unavailable, ignoring: %m", property);
-        else if (r < 0)
-               return log_device_debug_errno(dev, r, "Failed to read battery device property %s: %m", property);
-        else if (isempty(x))
-               log_device_debug(dev, "battery device property '%s' is null.", property);
-        else
-               siphash24_compress_string(x, state);
+        r = sd_device_get_sysattr_value(dev, attr, &x);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to read '%s' attribute: %m", attr);
+
+        if (!isempty(x))
+                siphash24_compress_string(x, state);
+
+        return 0;
+}
+
+static int siphash24_compress_id128(int (*getter)(sd_id128_t*), const char *name, struct siphash *state) {
+        sd_id128_t id;
+        int r;
+
+        assert(getter);
+        assert(state);
 
+        r = getter(&id);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get %s ID, ignoring: %m", name);
+
+        siphash24_compress(&id, sizeof(sd_id128_t), state);
         return 0;
 }
 
 /* Read system and battery identifier from specific location and generate hash of it */
 static int get_system_battery_identifier_hash(sd_device *dev, uint64_t *ret) {
         struct siphash state;
-        sd_id128_t machine_id, product_id;
-        int r;
 
         assert(ret);
         assert(dev);
 
         siphash24_init(&state, BATTERY_DISCHARGE_RATE_HASH_KEY.bytes);
 
-        get_battery_identifier(dev, "POWER_SUPPLY_MANUFACTURER", &state);
-        get_battery_identifier(dev, "POWER_SUPPLY_MODEL_NAME", &state);
-        get_battery_identifier(dev, "POWER_SUPPLY_SERIAL_NUMBER", &state);
-
-        r = sd_id128_get_machine(&machine_id);
-        if (r == -ENOENT)
-               log_debug_errno(r, "machine ID is unavailable: %m");
-        else if (r < 0)
-               return log_debug_errno(r, "Failed to get machine ID: %m");
-        else
-               siphash24_compress(&machine_id, sizeof(sd_id128_t), &state);
-
-        r = id128_get_product(&product_id);
-        if (r == -ENOENT)
-               log_debug_errno(r, "product_id does not exist: %m");
-        else if (r < 0)
-               return log_debug_errno(r, "Failed to get product ID: %m");
-        else
-               siphash24_compress(&product_id, sizeof(sd_id128_t), &state);
+        (void) siphash24_compress_device_sysattr(dev, "manufacturer", &state);
+        (void) siphash24_compress_device_sysattr(dev, "model_name", &state);
+        (void) siphash24_compress_device_sysattr(dev, "serial_number", &state);
+        (void) siphash24_compress_id128(sd_id128_get_machine, "machine", &state);
```

Comment:
only ENOENT should be ignored, but this now ignores every error, which doesn't seem right

Reply:
If we handle the failures in reading machine ID or product ID as critical, then we cannot load/save the discharge rates of all batteries on error. Especially for product ID, that is not controllable by users, as it is provided by firmware. So, if the firmware provides invalid ID, then we cannot use the estimated discharging rate. IIUC, adding the machine ID and product ID here is mostly for avoiding reuse of the calculated discharge rates on different systems. In usual setups, it is not necessary to consider such situation. Thus, the failure should not be handled as critical, but silently ignore the failure, except for the debugging logs.

Reply:
That's why ENOENT was skipped, but the rest was not - if there are permission denials or such, then it's a broken system, and it needs to be fixed. Hence please restore the checks like they were before - debug log and skip when not found, but propagate on other type of errors

Reply:
For machine ID, ENOENT also means the system is broken, no?? Then, why ENOENT is special?

For product ID, again, usually it is not controllable by users (except for the case people wrongly configures VM). Hence, refusing specific errors makes the feature cannot usable on machine with broken firmware. The product ID should be used as a hint, not strict requirement.

Reply:
whether it is present or not is not controllable yes, but other things are - for example, permission denials and so on - this risks making the hash unstable depending on transient issues (eg: selinux being changed to enforcing)